### DISCUSSION TITLE
Consumer App Portal read only mode

### DISCUSSION THREAD

**Discussion by dariom** [svix/svix-webhooks#1281] [Q&A] [answered]
Are you able to provide details on which features are/are not available in read only mode for the Consumer App Portal?

Unless I'm doing something wrong, in my testing I've found that in rea only mode I can:
* Create a new endpoint
* Edit an existing endpoint
* Delete an existing endpoint
* Disable/endale an endpoint
* Resend messages to an endpoint (either individually or in bulk)

Is that to be expected?

Is there any way in the Consumer App Portal to stop a user from modifying endpoints?

**Comment by tasn** [comment]
This shouldn't be the case. CC @svix-lucho to check this out.

**Comment by svix-lucho** [comment, marked as answer]
@dariom resending messages is allowed in the read-only mode of the App Portal. 

As for the edit/create operations on the endpoint, you should see something like this:
<img width="252" alt="Screenshot 2024-03-19 at 10 35 12 AM" src="https://github.com/svix/svix-webhooks/assets/133019767/41c27f1f-834b-4989-89c5-08e1100dd10c">

I just verified it and it seems to be working as expected. Can you make sure that you enabled read-only mode for the environment your app portal belongs to? Also, when making a change that affects the app portal, to see your change, you need to log in to the app portal from scratch again (using `app.svix.com/login#key=...`). Just refreshing the page is not enough.

**Reply by dariom** [reply]
Thanks @svix-lucho. I had simply been refreshing the portal, not logging in again. Can I suggest adding some additional guidance in your docs?

On a related note, I have some questions about white-labeling the App Portal. I see some elements of Svix content/branding/links in the custom portal. We'd hope to hide this once we move to a Professional plan - can you confirm whether this is the case?

Right now, I see "Svix" when loading the portal:
![image](https://github.com/svix/svix-webhooks/assets/384222/70568104-bd45-496e-8fc1-b37973505326)

in the bottom left-hand navigation:
![image](https://github.com/svix/svix-webhooks/assets/384222/77b13cfb-2184-4aeb-bd4a-0beaa8abc60e)

and after logging out (the custom logo isn't showing here either):
![image](https://github.com/svix/svix-webhooks/assets/384222/a9401f89-6186-444a-a15b-e16dd4f3c271)

Similarly, the help drop-down menu shows Svix-specific links:
![image](https://github.com/svix/svix-webhooks/assets/384222/1dbd46d7-ba90-4780-a0a6-776e79a14042)

Are there ways to address these to make the portal experience truly white-labeled and hide references to Svix?

I understand that the Event Catalog can be published using a custom domain. Is the same true of the App Portal?

**Reply by tasn** [reply]
@dariom, the app portal can also be fully whitelabeled in the free tier, you just need to embed it in your own UI which then removes all the Svix branding: https://docs.svix.com/app-portal#embedding-in-your-own-dashboard

**Reply by dariom** [reply]
Thanks,, we'll look into this. We were hoping not to create our own UI, but will explore this option.

**Comment by mderriey** [comment]
Hi,

Apologies for piggybacking on Dario's question, however it's closely related so I think it makes sense.

As we're exploring embedding the portal into one of our apps, we noticed that the URL we get back from the [Get Consumer App Portal Access endpoint](https://api.svix.com/docs#tag/Authentication/operation/v1.authentication.app-portal-access) contains a few query-string parameters, most specifically `readOnly=true`.

We found that if we remove this query string parameter, then the portal becomes editable &mdash; we've made sure to do these tests in incognito/InPrivate windows with no cookies or data in local storage.

It looks to us that given this is all happening client-side, it could mean someone could get access to an editable portal, which we want to avoid as we're enforcing limits through our API, like the maximum number of endpoints for an application.

I took a quick look, and it looks like we exchange the one-time token for a token; Would it be possible to look up the portal settings during this operation and issue a read-only token, so that even if the UI doesn't show the buttons disabled, the operations would fail when reaching the server?

I appreciate it's surely more complex than this, but I didn't want to come "empty-handed" to the conversation.

Thanks a lot.

**Reply by mderriey** [reply]
Friendly bump in case this went through the net with the Easter weekend 🙏

**Reply by tasn** [reply]
Yes, the app portal "read only" is aesthetic at the moment, but not actually backend enforced. That (along with other more fine-grained abilities) is coming soon.

**Reply by svix-jplatte** [reply]
App portal "read only" mode is now no longer aesthetic, as of a few weeks ago.
